src/codegeneration/ProperTailCallTransformer.js

```diff
@@ -0,0 +1,137 @@
+  createIdentifierExpression as id
+} from './ParseTreeFactory.js';
+import {
+  parseExpression,
+  parseStatement,
+  parseStatements
+} from './PlaceholderParser.js';
+import {
+  AnonBlock,
+  FunctionDeclaration,
+  FunctionExpression,
+} from '../syntax/trees/ParseTrees.js';
+
+/*
+ * Example:
+ *
+ * function f(a) {
+ *   var b = g(a);
+ *   f;
+ *   return h(b);
+ * }
+ *
+ * Becomes:
+ *
+ * $traceurRuntime.initTailRecursiveFunction(function f(a) {
+ *   return $traceurRuntime.call(function(a) {
+ *     b = g(a);
+ *     f;
+ *     return $traceurRuntime.continuation(h, null, [b]);
+ *   }, this, arguments);
+ * })
+ */
+
+export class ProperTailCallTransformer extends TempVarTransformer {
+  // TODO(mnieper): This transformer currently expects that classes and template
+  // literals have already been desugared. Otherwise they are not guaranteed
+  // to have proper tail calls.
+
+  constructor(identifierGenerator) {
+    super(identifierGenerator);
+    this.inBlock_ = false;
+  }
+
+  transformFunctionDeclaration(tree) {
+    var tree = super.transformFunctionDeclaration(tree);
+    if (tree.functionKind !== null) {
+      // do not transform async/generator functions
+      return tree;
+    }
+
+    var nameIdExpression = id(tree.name.identifierToken);
+
+    var setupFlagExpression = parseExpression
+        `$traceurRuntime.initTailRecursiveFunction(${nameIdExpression})`;
+
+    var funcDecl = this.transformFunction_(tree, FunctionDeclaration);
+    if (funcDecl === tree) {
+      return tree;
+    }
+
+    // Function declarations in blocks do not hoist. In that case we add the
+    // variable declaration after the function declaration.
+
+    var tmpVar = id(this.inBlock_ ?
+        this.getTempIdentifier() : this.addTempVar(setupFlagExpression));
+
+    if (!this.inBlock_) {
+      return funcDecl;
+    }
+
+    return new AnonBlock(null, [
+      funcDecl,
+      parseStatement `var ${tmpVar} = ${setupFlagExpression};`
+    ]);
+  }
+
+  transformFunctionExpression(tree) {
+    var tree = super.transformFunctionExpression(tree);
+    if (tree.functionKind) {
+      // do not transform async/generator functions
+      return tree;
+    }
+
+    var functionExpression =
+        this.transformFunction_(tree, FunctionExpression);
+
+    if (functionExpression === tree) {
+      return tree;
+    }
+
+    return parseExpression `
+        $traceurRuntime.initTailRecursiveFunction(${functionExpression})`;
+  }
+
+  transformFunction_(tree, constructor) {
+    var body = RewriteTailCallsTransformer.transform(this, tree.body);
+    if (body === tree.body) {
+      return tree;
+    }
+    var func = id(this.getTempIdentifier());
+    var parameterList = CloneTreeTransformer.cloneTree(tree.parameterList);
+    var innerFunction = createFunctionExpression(parameterList, body);
+    var outerBody = createFunctionBody(parseStatements `
+        return $traceurRuntime.call(${innerFunction}, this, arguments);`);
```

Comment:
This won't work because the inner function is not always called (by $traceurRuntime.call) with the same parameters as the wrapping function was called - nor do I see a need for this.